packages/grafana-ui/src/components/DateTimePickers/TimeRangeContext.tsx

```diff
@@ -0,0 +1,104 @@
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+
+import { TimeRange } from '@grafana/data';
+
+type TimeRangeContextValue = TimeRangeContextHookValue & {
+  // These are to be used internally and aren't passed to the users of the hook.
+
+  // Called when picker is mounted to update the picker count.
+  addPicker(): void;
+
+  // Called when picker is unmounted to update the picker count.
+  removePicker(): void;
+};
+
+export type TimeRangeContextHookValue = {
+  // If the time range is synced, this is the value that all pickers should show.
+  syncedValue?: TimeRange;
+
+  // If the time range is synced across all visible pickers.
+  synced: boolean;
+
+  // If it is possible to sync the time range. This is based just on the number of pickers that are visible so if
+  // there is only one, the sync button can be hidden.
+  syncPossible: boolean;
+
+  // Action passed to the picker to interact with the sync state.
+  // Sync the time range across all pickers with the provided value. Can be also used just to update a value when
+  // already synced.
+  sync(value: TimeRange): void;
+  unSync(): void;
+};
```

Comment:
Maybe move these to a separate `types` file?

Reply:
What would be the benefits? As I look at it, this is basically a return type of a function so in that sense I am not sure it makes sense to me to put it in a different file than the function itself.